<!DOCTYPE html>
<!-- This site was created with Hugo Blox. https://hugoblox.com -->
<!-- Last Published: July 31, 2025 --><html lang="en-us" >


<head><script src="/livereload.js?mindelay=10&amp;v=2&amp;port=1313&amp;path=livereload" data-no-instant defer></script>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  
  
  
    <meta name="generator" content="Hugo Blox Builder 5.9.7" />
  

  
  












  
  










  







  
  

  
  
  

  
  

  
  
    
    <script src="/js/mathjax-config.js"></script>
  

  

  <link rel="stylesheet" href="/css/vendor-bundle.min.css" media="print" onload="this.media='all'">

  
  
  
    
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/jpswalsh/academicons@1.9.4/css/academicons.min.css" integrity="sha512-IW0nhlW5MgNydsXJO40En2EoCkTTjZhI3yuODrZIc8cQ4h1XcF53PsqDHa09NqnkXuIe0Oiyyj171BqZFwISBw==" crossorigin="anonymous" media="print" onload="this.media='all'">
    

    
    
    
    
      
      
    
    
    

    
    
    
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/leaflet@1.7.1/dist/leaflet.min.css" integrity="" crossorigin="anonymous" media="print" onload="this.media='all'">
    

    

    
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
    
      
      

      
      

      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml.js" integrity="" crossorigin="anonymous" async></script>
      
    
      
      

      
      

      
    
      
      

      
      

      
    
  

  
  
  
  
  
  
  <link rel="stylesheet" href="/css/wowchemy.css" />

  
  
  

  
  
  
  
  
  
  
    
    
    <link rel="stylesheet" href="/css/libs/chroma/github-light.min.css" title="hl-light" media="print" onload="this.media='all'" >
    <link rel="stylesheet" href="/css/libs/chroma/dracula.min.css" title="hl-dark" media="print" onload="this.media='all'" disabled>
  

  
  



























  
  
  






  <meta name="author" content="Yang-Hong Dai 戴揚紘" />





  

<meta name="description" content="" />



<link rel="alternate" hreflang="en-us" href="http://localhost:1313/scrna_gem_flex_3/" />
<link rel="canonical" href="http://localhost:1313/scrna_gem_flex_3/" />



  <link rel="manifest" href="/manifest.webmanifest" />



<link rel="icon" type="image/png" href="/media/icon_hu7094183078009362424.png" />
<link rel="apple-touch-icon" type="image/png" href="/media/icon_hu6717257383916802474.png" />

<meta name="theme-color" content="#1565c0" />










  
  






<meta property="twitter:card" content="summary" />

  <meta property="twitter:site" content="@GetResearchDev" />
  <meta property="twitter:creator" content="@GetResearchDev" />
<meta property="twitter:image" content="http://localhost:1313/media/icon_hu4369633333135054090.png" />



  

<meta property="og:type" content="website" />
<meta property="og:site_name" content="BioMed Communications" />
<meta property="og:url" content="http://localhost:1313/scrna_gem_flex_3/" />
<meta property="og:title" content="單細胞RNA-seq數據分析教學-3：整合單細胞數據集 | BioMed Communications" />
<meta property="og:description" content="" /><meta property="og:image" content="http://localhost:1313/media/icon_hu4369633333135054090.png" /><meta property="og:locale" content="en-us" />

  
    <meta
      property="article:published_time"
      content="2025-04-14T00:00:00&#43;00:00"
    />
  
  
    <meta property="article:modified_time" content="2025-04-14T00:00:00&#43;00:00">
  







  




  
  
  

  
  

  


  
  <title>單細胞RNA-seq數據分析教學-3：整合單細胞數據集 | BioMed Communications</title>

  
  
  
  











</head>


<body id="top" data-spy="scroll" data-offset="70" data-target="#TableOfContents" class="page-wrapper   " data-wc-page-id="52bbf07a145d5fed33ba0a304657d310" >

  
  
  
  
  
  
  
  
  
  <script src="/js/wowchemy-init.min.js"></script>

  


<aside class="search-modal" id="search">
  <div class="container">
    <section class="search-header">

      <div class="row no-gutters justify-content-between mb-3">
        <div class="col-6">
          <h1>Search</h1>
        </div>
        <div class="col-6 col-search-close">
          <a class="js-search" href="#" aria-label="Close"><i class="fas fa-times-circle text-muted" aria-hidden="true"></i></a>
        </div>
      </div>

      <div id="search-box">
        
        <input name="q" id="search-query" placeholder="Search..." autocapitalize="off"
        autocomplete="off" autocorrect="off" spellcheck="false" type="search" class="form-control"
        aria-label="Search...">
        
      </div>

      
      

      

    </section>
    <section class="section-search-results">

      <div id="search-hits">
        
      </div>

    </section>
  </div>
</aside>



  <div class="page-header header--fixed">
  
  
  
  
  












<header>
  <nav class="navbar navbar-expand-lg navbar-light compensate-for-scrollbar" id="navbar-main">
    <div class="container-xl">

      
      <div class="d-none d-lg-inline-flex">
        <a class="navbar-brand" href="/">BioMed Communications</a>
      </div>
      

      
      <button type="button" class="navbar-toggler" data-toggle="collapse"
              data-target="#navbar-content" aria-controls="navbar-content" aria-expanded="false" aria-label="Toggle navigation">
      <span><i class="fas fa-bars"></i></span>
      </button>
      

      
      <div class="navbar-brand-mobile-wrapper d-inline-flex d-lg-none">
        <a class="navbar-brand" href="/">BioMed Communications</a>
      </div>
      

      
      
      <div class="navbar-collapse main-menu-item collapse justify-content-end" id="navbar-content">

        
        <ul class="navbar-nav d-md-inline-flex">
          

          

          
          
          

          

          
          
          
          

          
            
              
              
            
            
          

          <li class="nav-item">
            <a class="nav-link " href="/tour"><span>導覽</span></a>
          </li>

          
          

          

          
          
          

          

          
          
          
          

          
            
              
              
            
            
          

          <li class="nav-item">
            <a class="nav-link " href="/post"><span>文章</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
          

          <li class="nav-item">
            <a class="nav-link " href="/people"><span>人物</span></a>
          </li>

          
          

          

          
          
          
            
          

          

          
          
          
          

          
            
              
              
            
            
          

          <li class="nav-item">
            <a class="nav-link " href="/event"><span>消息</span></a>
          </li>

          
          

          

          
          
          

          

          
          
          
          

          
            
              
              
            
            
          

          <li class="nav-item">
            <a class="nav-link " href="/tech"><span>新技術分享</span></a>
          </li>

          
          

        

          
        </ul>
      </div>

      <ul class="nav-icons navbar-nav flex-row ml-auto d-flex pl-md-2">

        
        
          
        
          
        
          
        

        
        
        
        <li class="nav-item">
          <a class="nav-link js-search" href="#" aria-label="Search"><i class="fas fa-search" aria-hidden="true"></i></a>
        </li>
        

        
        
        

        
        

      </ul>

    </div>
  </nav>
</header>


  </div>

  <div class="page-body">
    
    
    

    <article class="article">

  













  

  
  
  
<div class="article-container pt-3">
  <h1>單細胞RNA-seq數據分析教學-3：整合單細胞數據集</h1>

  

  
    


<div class="article-metadata">

  
  
  
  
  <div>
    

  <span >
      <a href="/author/yang-hong-dai-%E6%88%B4%E6%8F%9A%E7%B4%98/">Yang-Hong Dai 戴揚紘</a></span>
  </div>
  
  

  

  

  

  
  
  
  
    <span class="middot-divider"></span>
    <a href="/scrna_gem_flex_3/#disqus_thread"></a>
  

  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fas fa-folder mr-1"></i><a href="/category/%E5%96%AE%E7%B4%B0%E8%83%9E%E5%AE%9A%E5%BA%8F/">單細胞定序</a></span>
  

</div>

    





  
</div>



  <div class="article-container">

    <div class="article-style">
      <h2 id="quick-look">Quick look</h2>
<p>在當前的研究實務中，很少只進行一次scRNA-seq實驗並僅產生一份scRNA-seq資料。這主要有兩個原因：首先，目前的scRNA-seq技術僅能在單次實驗中對有限的樣本提供訊息；若要跨實驗、多條件地測量多個樣本，通常需要將多次實驗所得的scRNA-seq資料進行聯合分析（Joint analysis）。儘管已有如<code>cell hashing</code>這樣的實驗策略，以及像<code>demuxlet</code>與<code>scSplit</code>這類可對樣本進行計算區分的方法 (是在同一次 scRNA-seq 實驗中混合多個樣本（例如不同病人的細胞）後，根據基因型或表現資料，將單細胞正確歸類到原始樣本身上)，在製備與定序scRNA-seq文庫時可以將多個樣本混合，但某些步驟（如組織分離, Tissue dissociation）仍然不可避免地需要針對不同樣本個別進行。因此，就如同在處理bulk RNA-seq資料時一樣，<code>批次效應（batch effect）</code>通常是分析中必須解決的一個關鍵干擾因素。</p>
<p>在本教學的此部分，我們將介紹幾種scRNA-seq資料整合方法。我們將使用範例教學中的DS1資料集與DS2資料集進行展示。值得注意的是，目前尚未有一種在所有情境下皆最優的整合方法，因此實務上重要的是嘗試不同的方法並比較它們的效果，最終選擇最適合該特定資料集的方法。</p>
<h2 id="載入資料">載入資料</h2>
<p>我們從匯入Seurat套件及儲存的Seurat物件開始：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">dplyr</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">patchwork</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">seurat_DS1</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&#34;DS1/seurat_obj_all.rds&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">seurat_DS2</span> <span class="o">&lt;-</span> <span class="nf">readRDS</span><span class="p">(</span><span class="s">&#34;DS2/seurat_obj_all.rds&#34;</span><span class="p">)</span>
</span></span></code></pre></div><h4 id="步驟1合併兩個資料集">步驟1：合併兩個資料集</h4>
<p>首先，有可能批次效應並不大，因而無需進行整合。因此我們應先觀察兩個資料集的表現，方式是先將它們合併：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">seurat</span> <span class="o">&lt;-</span> <span class="nf">merge</span><span class="p">(</span><span class="n">seurat_DS1</span><span class="p">,</span> <span class="n">seurat_DS2</span><span class="p">)</span> <span class="o">%&gt;%</span>
</span></span><span class="line"><span class="cl">  <span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">nfeatures</span> <span class="o">=</span> <span class="m">3000</span><span class="p">)</span> <span class="o">%&gt;%</span>
</span></span><span class="line"><span class="cl">  <span class="nf">ScaleData</span><span class="p">()</span> <span class="o">%&gt;%</span>
</span></span><span class="line"><span class="cl">  <span class="nf">RunPCA</span><span class="p">(</span><span class="n">npcs</span> <span class="o">=</span> <span class="m">50</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="c1">#需要先run PCA，然後用PCA的資訊喂給UMAP or TSNE</span>
</span></span><span class="line"><span class="cl">  <span class="nf">RunUMAP</span><span class="p">(</span><span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">plot1</span> <span class="o">&lt;-</span> <span class="nf">DimPlot</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">group.by</span><span class="o">=</span><span class="s">&#34;orig.ident&#34;</span><span class="p">)</span> <span class="c1">#用初始計畫名稱分類</span>
</span></span><span class="line"><span class="cl"><span class="n">plot2</span> <span class="o">&lt;-</span> <span class="nf">FeaturePlot</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;FOXG1&#34;</span><span class="p">,</span><span class="s">&#34;EMX1&#34;</span><span class="p">,</span><span class="s">&#34;DLX2&#34;</span><span class="p">,</span><span class="s">&#34;LHX9&#34;</span><span class="p">),</span> <span class="n">ncol</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">pt.size</span><span class="o">=</span><span class="m">0.1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">plot1</span> <span class="o">+</span> <span class="n">plot2</span> <span class="o">+</span> <span class="nf">plot_layout</span><span class="p">(</span><span class="n">widths</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1.5</span><span class="p">,</span> <span class="m">2</span><span class="p">))</span>
</span></span></code></pre></div><p>















<figure  id="figure-圖一">
  <div class="d-flex justify-content-center">
    <div class="w-100" ><img src="Fig1.png" alt="Fig1" loading="lazy" data-zoomable /></div>
  </div><figcaption>
      圖一
    </figcaption></figure>
</p>
<p>顯然，兩個資料集在UMAP嵌入圖上彼此分離。然而，marker基因的表現模式顯示，兩個資料集確實共用了相當多的細胞類型。理想情況下，相同細胞類型的細胞應該彼此混合。但由於批次效應，這並未發生，因此我們需要進行資料整合。整合的目標是讓相同類型的細胞能在兩個資料集中互相混合，而不同細胞類型或狀態的細胞則仍能區分。</p>
<ul>
<li>這邊merge的數據集應該要是已normalize但還未scale的數據。</li>
</ul>
<p>在這裡我們將嘗試以下幾種整合方法：</p>
<ol>
<li>
<p>Seurat</p>
</li>
<li>
<p>Harmony</p>
</li>
<li>
<p>LIGER</p>
</li>
<li>
<p>MNN</p>
</li>
</ol>
<h4 id="步驟2-1使用seurat進行資料整合">步驟2-1：使用Seurat進行資料整合</h4>
<p>Seurat 內建資料整合流程簡要來說，它首先對待整合的資料集使用 <code>Canonical Correlation Analysis（CCA）</code>，將它們各自旋轉，以最大化兩者的<code>共變化（canonical correlation）</code>。換句話說，Seurat 使用 CCA 尋找能<code>最大化資料集之間相似度的投影方式</code>。接著，Seurat引入<code>錨點（anchor）</code>機制，尋找在兩個資料集中的細胞配對錨點。這些錨點是指在CCA空間中互為最近鄰的細胞對，且彼此在各自資料集中的鄰居傾向於也彼此鄰近。這些被錨定的細胞被視為跨資料集的對應細胞，接著會計算<code>轉換矩陣</code>，並用於調整其中一個資料集的表達值。</p>
<p>欲使用Seurat進行整合，首先必須對每個待整合資料集進行正規化並找出高變異基因（若尚未完成）：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">seurat_DS1</span> <span class="o">&lt;-</span> <span class="nf">NormalizeData</span><span class="p">(</span><span class="n">seurat_DS1</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">nfeatures</span> <span class="o">=</span> <span class="m">3000</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">seurat_DS2</span> <span class="o">&lt;-</span> <span class="nf">NormalizeData</span><span class="p">(</span><span class="n">seurat_DS2</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">FindVariableFeatures</span><span class="p">(</span><span class="n">nfeatures</span> <span class="o">=</span> <span class="m">3000</span><span class="p">)</span>
</span></span></code></pre></div><p>接著，尋找資料集間的anchors。在這一步，Seurat接受一個Seurat物件清單作為輸入。請注意，Seurat支援超過兩個樣本的整合，只需將它們放入list中：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">seurat_objs</span> <span class="o">&lt;-</span> <span class="nf">list</span><span class="p">(</span><span class="n">DS1</span> <span class="o">=</span> <span class="n">seurat_DS1</span><span class="p">,</span> <span class="n">DS2</span> <span class="o">=</span> <span class="n">seurat_DS2</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">anchors</span> <span class="o">&lt;-</span> <span class="nf">FindIntegrationAnchors</span><span class="p">(</span><span class="n">object.list</span> <span class="o">=</span> <span class="n">seurat_objs</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">)</span>
</span></span></code></pre></div><ul>
<li>註：dims參數決定了要使用多少個CCA元件，可視需要微調。</li>
</ul>
<p>然後將這些anchor傳入<code>IntegrateData</code>函數以進行表達層次的修正：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">seurat</span> <span class="o">&lt;-</span> <span class="nf">IntegrateData</span><span class="p">(</span><span class="n">anchors</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">30</span><span class="p">)</span>
</span></span></code></pre></div><p>執行IntegrateData會產生一個新的Assay<code>（預設名稱為integrated）</code>，其中儲存經批次修正的表達矩陣。原始（未修正）表達值並未消失，而是保留在原本的Assay<code>（預設為RNA）</code>中。整合完成後，Seurat物件的預設Assay<code>會自動設為integrated</code>，但也可透過以下方式切換：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">seurat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s">&#34;RNA&#34;</span>
</span></span></code></pre></div><p>接著，使用整合後的Seurat物件進行後續分析，與第一部分的處理流程相同，只需略過正規化與高變基因找尋這兩步：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">seurat</span> <span class="o">&lt;-</span> <span class="nf">ScaleData</span><span class="p">(</span><span class="n">seurat</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">seurat</span> <span class="o">&lt;-</span> <span class="nf">RunPCA</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">npcs</span> <span class="o">=</span> <span class="m">50</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">seurat</span> <span class="o">&lt;-</span> <span class="nf">RunUMAP</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">seurat</span> <span class="o">&lt;-</span> <span class="nf">FindNeighbors</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">)</span> <span class="o">%&gt;%</span> <span class="nf">FindClusters</span><span class="p">(</span><span class="n">resolution</span> <span class="o">=</span> <span class="m">0.6</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">saveRDS</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s">&#34;integrated_seurat.rds&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>需要特別注意的是，雖然tSNE/UMAP嵌入與分群應使用整合後的Assay（integrated），但這些經修正的表達值不再適合作為定量分析依據。因此，進行群集標記基因鑑定或視覺化時，建議切換回原始Assay（RNA）：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">seurat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s">&#34;RNA&#34;</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="n">plot1</span> <span class="o">&lt;-</span> <span class="nf">UMAPPlot</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">group.by</span><span class="o">=</span><span class="s">&#34;orig.ident&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">plot2</span> <span class="o">&lt;-</span> <span class="nf">UMAPPlot</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">plot3</span> <span class="o">&lt;-</span> <span class="nf">FeaturePlot</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;FOXG1&#34;</span><span class="p">,</span><span class="s">&#34;EMX1&#34;</span><span class="p">,</span><span class="s">&#34;DLX2&#34;</span><span class="p">,</span><span class="s">&#34;LHX9&#34;</span><span class="p">),</span> <span class="n">ncol</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">pt.size</span><span class="o">=</span><span class="m">0.1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">((</span><span class="n">plot1</span> <span class="o">/</span> <span class="n">plot2</span><span class="p">)</span> <span class="o">|</span> <span class="n">plot3</span><span class="p">)</span> <span class="o">+</span> <span class="nf">plot_layout</span><span class="p">(</span><span class="n">width</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">))</span>
</span></span></code></pre></div><p>















<figure  id="figure-圖二">
  <div class="d-flex justify-content-center">
    <div class="w-100" ><img src="Fig2.png" alt="Fig2" loading="lazy" data-zoomable /></div>
  </div><figcaption>
      圖二
    </figcaption></figure>
</p>
<p>雖不完美，但這樣已能使兩個資料集更具可比性。</p>
<p>若想進一步優化結果，可調參數包括：</p>
<ul>
<li>
<p>FindVariableFeatures中的nfeatures參數：影響整合用基因集的選擇。</p>
</li>
<li>
<p>FindIntegrationAnchors中的<code>anchor.features</code>、<code>dims</code>、<code>k.anchor</code>、<code>k.filter</code>、<code>k.score</code>等。</p>
</li>
<li>
<p>IntegrateData中的dims參數。</p>
</li>
</ul>
<h4 id="步驟2-2使用harmony進行資料整合">步驟2-2：使用Harmony進行資料整合</h4>
<p>除了Seurat，現在已有更多可用的資料整合方法。<code>Harmony</code>就是其中一種，由Soumya Raychaudhuri實驗室開發。它是第一個系統性評估scRNA-seq整合方法的基準研究中最受矚目的方法之一。</p>
<h4 id="harmony的原理簡介">Harmony的原理簡介：</h4>
<p>Harmony使用<code>模糊分群</code>將每個細胞分配至多個群集中，對於每個群集，它計算各資料集需移動其該群集的質心至全域群集質心所需的修正量。由於每個細胞屬於<code>多個群集</code>，因此細胞的修正量是這些<code>群集修正量的加權平均</code>。這個過程會重複進行直到收斂或達到最大迭代次數。</p>
<ol>
<li>Harmony 在已經 PCA 降維過的空間上進行（例如前 20 個 principal components）。</li>
<li>Harmony 將資料初始化為數個「暫時性群集」。這些群集依照 PCA 空間距離分配，並記錄每個細胞所屬的 batch ID（例如 sample A、B、C）</li>
<li>每一個迭代中：根據每個 cluster 裡不同 batch 的分布，計算 batch-specific centroid。將每個細胞的座標往一個「均衡位置」移動（weighted correction）</li>
</ol>
<ul>
<li>目標是：讓同一個 cluster 裡的細胞，不論來自哪個 batch，都更接近 cluster 中心</li>
<li>更新細胞對 cluster 的隸屬機率（soft assignment）</li>
<li>每個細胞可以同時屬於多個 cluster（機率形式），非硬性分群 → 重複這個過程數輪，直到收斂</li>
</ul>
<ol start="4">
<li>最終輸出是 批次已修正的 PCA 向量（通常儲存在 harmony_embeddings)。
這個新空間可拿來做：</li>
</ol>
<ul>
<li>RunUMAP() → 畫 UMAP</li>
<li>FindNeighbors() → 建圖</li>
<li>FindClusters() → 分群</li>
</ul>
<p>Harmony提供一個可直接套用至Seurat物件的函數：<code>RunHarmony</code>。它接受先前合併的Seurat物件，並須指定哪一欄metadata作為批次來源：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">harmony</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">seurat</span> <span class="o">&lt;-</span> <span class="nf">RunHarmony</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">group.by.vars</span> <span class="o">=</span> <span class="s">&#34;orig.ident&#34;</span><span class="p">,</span> <span class="n">dims.use</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">,</span> <span class="n">max.iter.harmony</span> <span class="o">=</span> <span class="m">50</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">seurat</span> <span class="o">&lt;-</span> <span class="nf">RunUMAP</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">reduction</span> <span class="o">=</span> <span class="s">&#34;harmony&#34;</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">seurat</span> <span class="o">&lt;-</span> <span class="nf">FindNeighbors</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">reduction</span> <span class="o">=</span> <span class="s">&#34;harmony&#34;</span><span class="p">,</span> <span class="n">dims</span> <span class="o">=</span> <span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">)</span> <span class="o">%&gt;%</span>
</span></span><span class="line"><span class="cl">  <span class="nf">FindClusters</span><span class="p">(</span><span class="n">resolution</span> <span class="o">=</span> <span class="m">0.6</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">saveRDS</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">file</span><span class="o">=</span><span class="s">&#34;integrated_harmony.rds&#34;</span><span class="p">)</span>
</span></span></code></pre></div><ul>
<li>
<p>dims.use設定要使用的PCA維度。</p>
</li>
<li>
<p>max.iter.harmony設定最大迭代次數。</p>
</li>
</ul>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">plot1</span> <span class="o">&lt;-</span> <span class="nf">UMAPPlot</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">group.by</span><span class="o">=</span><span class="s">&#34;orig.ident&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">plot2</span> <span class="o">&lt;-</span> <span class="nf">UMAPPlot</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">label</span> <span class="o">=</span> <span class="bp">T</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">plot3</span> <span class="o">&lt;-</span> <span class="nf">FeaturePlot</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;FOXG1&#34;</span><span class="p">,</span><span class="s">&#34;EMX1&#34;</span><span class="p">,</span><span class="s">&#34;DLX2&#34;</span><span class="p">,</span><span class="s">&#34;LHX9&#34;</span><span class="p">),</span> <span class="n">ncol</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">pt.size</span><span class="o">=</span><span class="m">0.1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="p">((</span><span class="n">plot1</span> <span class="o">/</span> <span class="n">plot2</span><span class="p">)</span> <span class="o">|</span> <span class="n">plot3</span><span class="p">)</span> <span class="o">+</span> <span class="nf">plot_layout</span><span class="p">(</span><span class="n">width</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">))</span>
</span></span></code></pre></div><p>















<figure  id="figure-圖三">
  <div class="d-flex justify-content-center">
    <div class="w-100" ><img src="Fig3.png" alt="Fig3" loading="lazy" data-zoomable /></div>
  </div><figcaption>
      圖三
    </figcaption></figure>
</p>
<p>整合效果不錯，兩樣本的細胞彼此良好混合，並能觀察到一些發育軌跡。不過，<code>對某些混合的群體</code>（特別是非dorsal telencephalic細胞）仍需進一步判斷它們是否確為相同細胞類型。</p>
<p>由於Harmony<code>以PCA結果作為輸入</code>，並於PCA空間內進行修正，因此影響PCA的參數（如高變基因選取的nfeatures）亦會影響整合結果。RunHarmony<code>預設會使用所有計算出來的主成分</code>，但可透過設定dims.use加以指定。</p>
<p>需要注意的是，和CCA類似，如果細胞具有「生物上相似的表現特徵」，無論來自哪個 sample，它們會被聚在一起，弭平batch effect。但如果兩個samples是來自不同實驗條件，這樣的整合反而會干擾真正的生物變異，需要特別注意。</p>
<h4 id="步驟2-3使用liger進行資料整合">步驟2-3：使用LIGER進行資料整合</h4>
<p>LIGER（Linked Inference of Genomic Experimental Relationships）是一種基於<code>非負矩陣分解（Non-negative Matrix Factorization, NMF）</code>的整合方法。它將每個資料集分解為共同與特有的因子矩陣 (NMF精髓)，藉此找出跨資料集的共享細胞結構，同時保留每個資料集的獨特性。</p>
<p>LIGER整合步驟概要：</p>
<ol>
<li>
<p>正規化資料集（通常以log-normalization處理）</p>
</li>
<li>
<p>將資料轉換為LIGER所需的格式（通常為liger::createLiger函數）</p>
</li>
<li>
<p>執行整合，包括optimizeALS()與quantile_norm()兩個關鍵步驟</p>
</li>
<li>
<p>使用整合後的空間做可視化與分群</p>
</li>
</ol>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">liger</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># 將Seurat物件轉為list格式並取出表達矩陣</span>
</span></span><span class="line"><span class="cl"><span class="n">counts1</span> <span class="o">&lt;-</span> <span class="nf">GetAssayData</span><span class="p">(</span><span class="n">seurat_DS1</span><span class="p">,</span> <span class="n">slot</span> <span class="o">=</span> <span class="s">&#34;counts&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">counts2</span> <span class="o">&lt;-</span> <span class="nf">GetAssayData</span><span class="p">(</span><span class="n">seurat_DS2</span><span class="p">,</span> <span class="n">slot</span> <span class="o">=</span> <span class="s">&#34;counts&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># 建立LIGER物件</span>
</span></span><span class="line"><span class="cl"><span class="n">liger_obj</span> <span class="o">&lt;-</span> <span class="nf">createLiger</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">DS1</span> <span class="o">=</span> <span class="n">counts1</span><span class="p">,</span> <span class="n">DS2</span> <span class="o">=</span> <span class="n">counts2</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="n">liger_obj</span> <span class="o">&lt;-</span> <span class="nf">normalize</span><span class="p">(</span><span class="n">liger_obj</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">liger_obj</span> <span class="o">&lt;-</span> <span class="nf">selectGenes</span><span class="p">(</span><span class="n">liger_obj</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">liger_obj</span> <span class="o">&lt;-</span> <span class="nf">scaleNotCenter</span><span class="p">(</span><span class="n">liger_obj</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">liger_obj</span> <span class="o">&lt;-</span> <span class="nf">optimizeALS</span><span class="p">(</span><span class="n">liger_obj</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="m">20</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">liger_obj</span> <span class="o">&lt;-</span> <span class="nf">quantile_norm</span><span class="p">(</span><span class="n">liger_obj</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># 可視化與群集分析</span>
</span></span><span class="line"><span class="cl"><span class="n">liger_obj</span> <span class="o">&lt;-</span> <span class="nf">runUMAP</span><span class="p">(</span><span class="n">liger_obj</span><span class="p">,</span> <span class="n">distance</span> <span class="o">=</span> <span class="s">&#39;cosine&#39;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">liger_obj</span> <span class="o">&lt;-</span> <span class="nf">findClusters</span><span class="p">(</span><span class="n">liger_obj</span><span class="p">,</span> <span class="n">resolution</span> <span class="o">=</span> <span class="m">0.4</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">plotByDatasetAndCluster</span><span class="p">(</span><span class="n">liger_obj</span><span class="p">)</span>
</span></span></code></pre></div><p>















<figure  id="figure-圖四">
  <div class="d-flex justify-content-center">
    <div class="w-100" ><img src="Fig4.png" alt="Fig4" loading="lazy" data-zoomable /></div>
  </div><figcaption>
      圖四
    </figcaption></figure>

注意事項：</p>
<ul>
<li>
<p>LIGER 偏好<code>使用原始counts資料</code>，並自行處理正規化與篩選步驟。</p>
</li>
<li>
<p><code>k</code>為潛在因子的數目，是可調參數，應根據資料集大小與異質性設定。</p>
</li>
</ul>
<h4 id="步驟2-4使用mnnmutual-nearest-neighbors進行整合">步驟2-4：使用MNN（Mutual Nearest Neighbors）進行整合</h4>
<p>MNN整合方法由Haghverdi等人提出，核心概念是尋找在不同資料集中<code>互為最近鄰的細胞對</code>，藉此計算校正向量並調整批次間的偏移。此方法由<code>batchelo</code>r與<code>scran</code>套件實作。</p>
<p>MNN整合步驟：</p>
<ol>
<li>
<p>將Seurat物件轉為SingleCellExperiment（SCE）格式</p>
</li>
<li>
<p>使用fastMNN()計算整合表達空間</p>
</li>
<li>
<p>將結果用於後續降維與分群</p>
</li>
</ol>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">Seurat</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">batchelor</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">SingleCellExperiment</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">scater</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># 轉換為SCE格式</span>
</span></span><span class="line"><span class="cl"><span class="n">sce1</span> <span class="o">&lt;-</span> <span class="nf">as.SingleCellExperiment</span><span class="p">(</span><span class="n">seurat_DS1</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">sce2</span> <span class="o">&lt;-</span> <span class="nf">as.SingleCellExperiment</span><span class="p">(</span><span class="n">seurat_DS2</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># 使用MNN整合</span>
</span></span><span class="line"><span class="cl"><span class="n">sce.mnn</span> <span class="o">&lt;-</span> <span class="nf">fastMNN</span><span class="p">(</span><span class="n">sce1</span><span class="p">,</span> <span class="n">sce2</span><span class="p">,</span> <span class="n">d</span> <span class="o">=</span> <span class="m">50</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="m">20</span><span class="p">)</span>
</span></span><span class="line"><span class="cl">
</span></span><span class="line"><span class="cl"><span class="c1"># 可視化與分群</span>
</span></span><span class="line"><span class="cl"><span class="n">sce.mnn</span> <span class="o">&lt;-</span> <span class="nf">runUMAP</span><span class="p">(</span><span class="n">sce.mnn</span><span class="p">,</span> <span class="n">dimred</span> <span class="o">=</span> <span class="s">&#34;corrected&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">plotUMAP</span><span class="p">(</span><span class="n">sce.mnn</span><span class="p">,</span> <span class="n">colour_by</span> <span class="o">=</span> <span class="s">&#34;batch&#34;</span><span class="p">)</span>
</span></span></code></pre></div><p>















<figure  id="figure-圖五">
  <div class="d-flex justify-content-center">
    <div class="w-100" ><img src="Fig5.png" alt="Fig5" loading="lazy" data-zoomable /></div>
  </div><figcaption>
      圖五
    </figcaption></figure>

注意事項：</p>
<ul>
<li>
<p>MNN需要先對每個資料集進行正規化（通常建議log-normalization）</p>
</li>
<li>
<p>d為PCA維度數，k為尋找MNN時的鄰近細胞數目</p>
</li>
<li>
<p>整合後的降維結果儲存在corrected中</p>
</li>
</ul>
<hr>
<h2 id="小結與建議">小結與建議</h2>
<p>本課程至此介紹了四種主流的scRNA-seq資料整合方法，包括：</p>
<ol>
<li>
<p>Seurat（使用CCA與Anchoring機制）</p>
</li>
<li>
<p>Harmony（基於PCA的模糊分群與向量校正）</p>
</li>
<li>
<p>LIGER（基於NMF的共享與專屬因子分析）</p>
</li>
<li>
<p>MNN（尋找跨批次互為最近鄰的細胞進行修正）</p>
</li>
</ol>
<p>這些方法各有優缺點，適用於不同情境，實作時建議：</p>
<ol>
<li>
<p>若以Seurat為主流程，Seurat與Harmony較為便利整合</p>
</li>
<li>
<p>LIGER較適合於細胞異質性高或跨物種分析</p>
</li>
<li>
<p>MNN理論簡單，適用於數量不多但變異大的樣本</p>
</li>
</ol>
<hr>
<p>下一步，我們將示範如何比較整合後的結果、檢查整合品質（如kBET、LISI等評估指標），並進行下游分析（如差異表達、軌跡分析等）。</p>
<h4 id="步驟3整合品質評估與比較">步驟3：整合品質評估與比較</h4>
<p>整合的最終目的是讓相同細胞類型在不同樣本之間能夠準確對齊，同時保留真實的生物變異。因此，我們需要使用客觀指標來評估整合結果的品質。以下介紹幾個常見且被廣泛使用的指標：</p>
<h4 id="31-kbetk-nearest-neighbor-batch-effect-test">3.1 kBET（k-nearest neighbor Batch Effect Test）</h4>
<p>kBET 是一種基於最近鄰的統計測試方法，用來檢測在局部區域內不同批次細胞是否混合良好。理想情況下，不同批次細胞應在嵌入空間中彼此混合，若存在顯著的偏離，則說明仍存在批次效應。</p>
<p>可使用 <code>kBET R 套件</code>或 <code>scIB（Python）</code>來進行測試。</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="c1"># 使用kBET評估</span>
</span></span><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">kBET</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">kbet_result</span> <span class="o">&lt;-</span> <span class="nf">kBET</span><span class="p">(</span><span class="n">data_matrix</span><span class="p">,</span> <span class="n">batch_labels</span><span class="p">,</span> <span class="n">k</span> <span class="o">=</span> <span class="m">30</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">summary</span><span class="p">(</span><span class="n">kbet_result</span><span class="o">$</span><span class="n">stats</span><span class="p">)</span>
</span></span></code></pre></div><p>指標：Rejection rate 越低表示混合越好（接近 0 為佳）。</p>
<h4 id="32-lisilocal-inverse-simpsons-index">3.2 LISI（Local Inverse Simpson&rsquo;s Index）</h4>
<p>LISI 衡量的是局部鄰域的多樣性，分為：</p>
<ol>
<li>
<p>iLISI（Integration LISI）：評估批次混合程度</p>
</li>
<li>
<p>cLISI（Clustering LISI）：評估細胞類型區分能力</p>
</li>
</ol>
<p>可使用 lisi 套件：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">lisi</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">lisi_scores</span> <span class="o">&lt;-</span> <span class="nf">compute_lisi</span><span class="p">(</span><span class="n">embedding</span><span class="p">,</span> <span class="n">metadata</span><span class="p">,</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;batch&#34;</span><span class="p">,</span> <span class="s">&#34;cell_type&#34;</span><span class="p">))</span>
</span></span></code></pre></div><ul>
<li>
<p>iLISI 越高越好（表示批次混合良好）</p>
</li>
<li>
<p>cLISI 越低越好（表示細胞類型分離明確）</p>
</li>
</ul>
<h4 id="33-silhouette-score">3.3 Silhouette Score</h4>
<p>Silhouette Score 衡量細胞與其同群細胞間的距離是否小於與其他群細胞的距離。常用於評估群集清晰度。</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">cluster</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">sil_scores</span> <span class="o">&lt;-</span> <span class="nf">silhouette</span><span class="p">(</span><span class="n">cluster_labels</span><span class="p">,</span> <span class="nf">dist</span><span class="p">(</span><span class="n">embedding</span><span class="p">))</span>
</span></span><span class="line"><span class="cl"><span class="nf">summary</span><span class="p">(</span><span class="n">sil_scores</span><span class="p">)</span>
</span></span></code></pre></div><p>分數範圍 -1 到 1，越高表示群集越分明。</p>
<h4 id="步驟4下游分析與應用">步驟4：下游分析與應用</h4>
<p>整合完成後，即可進行多樣的下游分析。以下列出幾個常見應用：</p>
<h4 id="41-群集標記基因鑑定cluster-marker-identification">4.1 群集標記基因鑑定（Cluster Marker Identification）</h4>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">DefaultAssay</span><span class="p">(</span><span class="n">seurat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s">&#34;RNA&#34;</span>
</span></span><span class="line"><span class="cl"><span class="n">markers</span> <span class="o">&lt;-</span> <span class="nf">FindAllMarkers</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">only.pos</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">min.pct</span> <span class="o">=</span> <span class="m">0.25</span><span class="p">,</span> <span class="n">logfc.threshold</span> <span class="o">=</span> <span class="m">0.25</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">head</span><span class="p">(</span><span class="n">markers</span><span class="p">)</span>
</span></span></code></pre></div><h4 id="42-差異表達分析differential-expression-analysis">4.2 差異表達分析（Differential Expression Analysis）</h4>
<p>可用於比較不同條件或群集之間的基因表現：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="n">de_genes</span> <span class="o">&lt;-</span> <span class="nf">FindMarkers</span><span class="p">(</span><span class="n">seurat</span><span class="p">,</span> <span class="n">ident.1</span> <span class="o">=</span> <span class="s">&#34;Cluster1&#34;</span><span class="p">,</span> <span class="n">ident.2</span> <span class="o">=</span> <span class="s">&#34;Cluster2&#34;</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">head</span><span class="p">(</span><span class="n">de_genes</span><span class="p">)</span>
</span></span></code></pre></div><h4 id="43-發育軌跡分析trajectory--pseudotime-analysis">4.3 發育軌跡分析（Trajectory / Pseudotime Analysis）</h4>
<p>可使用 Monocle3、Slingshot 等工具建立細胞狀態轉變軌跡：</p>
<div class="highlight"><pre tabindex="0" class="chroma"><code class="language-r" data-lang="r"><span class="line"><span class="cl"><span class="nf">library</span><span class="p">(</span><span class="n">monocle3</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">as.cell_data_set</span><span class="p">(</span><span class="n">seurat</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">cluster_cells</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">learn_graph</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="n">cds</span> <span class="o">&lt;-</span> <span class="nf">order_cells</span><span class="p">(</span><span class="n">cds</span><span class="p">)</span>
</span></span><span class="line"><span class="cl"><span class="nf">plot_cells</span><span class="p">(</span><span class="n">cds</span><span class="p">,</span> <span class="n">color_cells_by</span> <span class="o">=</span> <span class="s">&#34;pseudotime&#34;</span><span class="p">)</span>
</span></span></code></pre></div><hr>
<h2 id="課程小結">課程小結</h2>
<p>本課程從scRNA-seq整合的背景與必要性出發，依序介紹了四種整合工具（Seurat, Harmony, LIGER, MNN），並補充如何進行整合品質評估（kBET, LISI, Silhouette Score）以及典型的下游應用分析。透過這些工具與方法，研究者可以更準確解讀跨樣本、多條件的單細胞資料，為後續生物學發現與機制探討奠定基礎。</p>
    </div>

    





<div class="article-tags">
  
  <a class="badge badge-light" href="/tag/%E5%9F%BA%E5%9B%A0%E9%AB%94%E5%AD%B8/">基因體學</a>
  
  <a class="badge badge-light" href="/tag/%E7%94%9F%E7%89%A9%E8%B3%87%E8%A8%8A%E5%AD%B8/">生物資訊學</a>
  
</div>



<div class="share-box">
  <ul class="share">
    
      
      
      
        
      
      
      
      
      
      
      
      <li>
        <a href="https://twitter.com/intent/tweet?url=http%3A%2F%2Flocalhost%3A1313%2Fscrna_gem_flex_3%2F&amp;text=%E5%96%AE%E7%B4%B0%E8%83%9ERNA-seq%E6%95%B8%E6%93%9A%E5%88%86%E6%9E%90%E6%95%99%E5%AD%B8-3%EF%BC%9A%E6%95%B4%E5%90%88%E5%96%AE%E7%B4%B0%E8%83%9E%E6%95%B8%E6%93%9A%E9%9B%86" target="_blank" rel="noopener" class="share-btn-twitter" aria-label="twitter">
          <i class="fab fa-twitter"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      
      
      
      
      <li>
        <a href="https://www.facebook.com/sharer.php?u=http%3A%2F%2Flocalhost%3A1313%2Fscrna_gem_flex_3%2F&amp;t=%E5%96%AE%E7%B4%B0%E8%83%9ERNA-seq%E6%95%B8%E6%93%9A%E5%88%86%E6%9E%90%E6%95%99%E5%AD%B8-3%EF%BC%9A%E6%95%B4%E5%90%88%E5%96%AE%E7%B4%B0%E8%83%9E%E6%95%B8%E6%93%9A%E9%9B%86" target="_blank" rel="noopener" class="share-btn-facebook" aria-label="facebook">
          <i class="fab fa-facebook"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      
      
      
      
        
      
      <li>
        <a href="mailto:?subject=%E5%96%AE%E7%B4%B0%E8%83%9ERNA-seq%E6%95%B8%E6%93%9A%E5%88%86%E6%9E%90%E6%95%99%E5%AD%B8-3%EF%BC%9A%E6%95%B4%E5%90%88%E5%96%AE%E7%B4%B0%E8%83%9E%E6%95%B8%E6%93%9A%E9%9B%86&amp;body=http%3A%2F%2Flocalhost%3A1313%2Fscrna_gem_flex_3%2F" target="_blank" rel="noopener" class="share-btn-email" aria-label="envelope">
          <i class="fas fa-envelope"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      
      
      
      
      <li>
        <a href="https://www.linkedin.com/shareArticle?url=http%3A%2F%2Flocalhost%3A1313%2Fscrna_gem_flex_3%2F&amp;title=%E5%96%AE%E7%B4%B0%E8%83%9ERNA-seq%E6%95%B8%E6%93%9A%E5%88%86%E6%9E%90%E6%95%99%E5%AD%B8-3%EF%BC%9A%E6%95%B4%E5%90%88%E5%96%AE%E7%B4%B0%E8%83%9E%E6%95%B8%E6%93%9A%E9%9B%86" target="_blank" rel="noopener" class="share-btn-linkedin" aria-label="linkedin-in">
          <i class="fab fa-linkedin-in"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      
      
      
      
      <li>
        <a href="whatsapp://send?text=%E5%96%AE%E7%B4%B0%E8%83%9ERNA-seq%E6%95%B8%E6%93%9A%E5%88%86%E6%9E%90%E6%95%99%E5%AD%B8-3%EF%BC%9A%E6%95%B4%E5%90%88%E5%96%AE%E7%B4%B0%E8%83%9E%E6%95%B8%E6%93%9A%E9%9B%86%20http%3A%2F%2Flocalhost%3A1313%2Fscrna_gem_flex_3%2F" target="_blank" rel="noopener" class="share-btn-whatsapp" aria-label="whatsapp">
          <i class="fab fa-whatsapp"></i>
        </a>
      </li>
    
      
      
      
        
      
      
      
      
      
      
      
      <li>
        <a href="https://service.weibo.com/share/share.php?url=http%3A%2F%2Flocalhost%3A1313%2Fscrna_gem_flex_3%2F&amp;title=%E5%96%AE%E7%B4%B0%E8%83%9ERNA-seq%E6%95%B8%E6%93%9A%E5%88%86%E6%9E%90%E6%95%99%E5%AD%B8-3%EF%BC%9A%E6%95%B4%E5%90%88%E5%96%AE%E7%B4%B0%E8%83%9E%E6%95%B8%E6%93%9A%E9%9B%86" target="_blank" rel="noopener" class="share-btn-weibo" aria-label="weibo">
          <i class="fab fa-weibo"></i>
        </a>
      </li>
    
  </ul>
</div>










  
  
  

  

  
  <section id="comments">
    
<div id="disqus_thread"></div>
<script>
  var disqus_config = function () {
    
    
    
  };
  (function() {
    if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
      document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
      return;
    }
    var d = document, s = d.createElement('script'); s.async = true;
    s.src = 'https://' + "yhd" + '.disqus.com/embed.js';
    s.setAttribute('data-timestamp', +new Date());
    (d.head || d.body).appendChild(s);
  })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>


  </section>
  










  </div>
</article>
  </div>

  <div class="page-footer">
    
    
    <div class="container">
      <footer class="site-footer">

  












  
  
  
  
  













  
  
  

  
  
    
  
  
    
  

  

  
  <p class="powered-by copyright-license-text">
    © 2025. This work is licensed under <a href="https://creativecommons.org/licenses/by-nc-nd/4.0" rel="noopener noreferrer" target="_blank">CC BY NC ND 4.0</a>
  </p>
  

  <p class="powered-by footer-license-icons">
    <a href="https://creativecommons.org/licenses/by-nc-nd/4.0" rel="noopener noreferrer" target="_blank" aria-label="Creative Commons">
      <i class="fab fa-creative-commons fa-2x" aria-hidden="true"></i>
      <i class="fab fa-creative-commons-by fa-2x" aria-hidden="true"></i>
      
        <i class="fab fa-creative-commons-nc fa-2x" aria-hidden="true"></i>
      
      
        <i class="fab fa-creative-commons-nd fa-2x" aria-hidden="true"></i>
      
    </a>
  </p>





  <p class="powered-by">
    
    
    
      
      
      
      
      
      
      Published with <a href="https://hugoblox.com/?utm_campaign=poweredby" target="_blank" rel="noopener">Hugo Blox Builder</a> — the free, <a href="https://github.com/HugoBlox/hugo-blox-builder" target="_blank" rel="noopener">open source</a> website builder that empowers creators.
    
  </p>
</footer>

    </div>
    
  </div>

  


<script src="/js/vendor-bundle.min.js"></script>




  

  
  

  






  <script src="https://cdn.jsdelivr.net/npm/leaflet@1.7.1/dist/leaflet.min.js" integrity="" crossorigin="anonymous"></script>








  
  <script id="search-hit-fuse-template" type="text/x-template">
    <div class="search-hit" id="summary-{{key}}">
      <div class="search-hit-content">
        <div class="search-hit-name">
          <a href="{{relpermalink}}">{{title}}</a>
          <div class="article-metadata search-hit-type">{{type}}</div>
          <p class="search-hit-description">{{snippet}}</p>
        </div>
      </div>
    </div>
  </script>
  
    <script src="https://cdn.jsdelivr.net/gh/krisk/Fuse@v3.2.1/dist/fuse.min.js" integrity="sha512-o38bmzBGX+hD3JHWUFCDA09btWaqrNmoJ3RXLlrysA7PP01Kgs4UlE4MhelE1v5dJR3+cxlR4qQlotsW7jKsnw==" crossorigin="anonymous"></script>
    <script src="https://cdn.jsdelivr.net/gh/julmot/mark.js@8.11.1/dist/jquery.mark.min.js" integrity="sha512-mhbv5DqBMgrWL+32MmsDOt/OAvqr/cHimk6B8y/bx/xS88MVkYGPiVv2ixKVrkywF2qHplNRUvFsAHUdxZ3Krg==" crossorigin="anonymous"></script>
  









  <script id="dsq-count-scr" src="https://yhd.disqus.com/count.js" async></script>




  
  
  
  
  
  
  

















<script id="page-data" type="application/json">{"use_headroom":true}</script>


  <script src="/js/wowchemy-headroom.js" type="module"></script>









  
  


<script src="/en/js/wowchemy.min.js"></script>



  <script src="/js/wowchemy-map.js" type="module"></script>




  
<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <h5 class="modal-title">Cite</h5>
        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
          <span aria-hidden="true">&times;</span>
        </button>
      </div>
      <div class="modal-body">
        
        <pre><code></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-outline-primary my-1 js-copy-cite" href="#" target="_blank">
          <i class="fas fa-copy"></i> Copy
        </a>
        <a class="btn btn-outline-primary my-1 js-download-cite" href="#" target="_blank">
          <i class="fas fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>


  <script src="/js/wowchemy-publication.js" type="module"></script>


















</body>
</html>
